ospfd: OSPFv2 VRF Support

chiragshah6 opened the pull request:
This is first cut of OSFv2 VRF Support.

struct ospf cotains vrf_name and vrf_id along with instance
ospf master will have number of struct ospf VRF aware instances.

A new ospf_lookup API can take instance and vrf_id as input. Multi-instance and VRF support is not tested/claimed.
ospf_lookup () is the legacy lookup api which would return default instance. Along with this lookby vrf_id and lookup by instance_and_vrf_id apis are implemented.

Default instance (assigend VRF_DEFAULT) is created when ospf daemon is spawned.

ospf master iflist only contains for default vrf interface list. Instead vrf interface list is used to traverse per VRF interface set.

List of ospf interface and list of area are part of "struct ospf" which would devide per vrf.

LSA and LSDB would be operate on per AREA and per ospf interface which would use its parent ospf vrf_id.

OSPF Route Add/delete will have struct ospf or vrf_id passed as argument and ospf->vrf_id used to program the routes to zebra.

router ospf vrf vrf1012 configured if vrf is not active, oi->running will be false so router_id update is skipped and called in vrf_enabled callback function.

Primary OSPF show CLIs like neighbor, interface, database, route are converted to take vrf <all | word> as arguement and follows "show ip ospf vrf [all|word] ....." format.
json output containing VRF anchor point will be added in future check-ins.

A new CLI 'show ip ospf vrfs' outlines all vrf configured ospf instances.

A new OSPF debug trace is added and used for internal and non critical daemon debugging activity

 OSPF SNMP, Router-Information, MPLE-TE, Opaque-LSA  are not support by VRF.

Signed-off-by: Chirag Shah <chirag@cumulusnetworks.com>

donaldsharp commented on ospfd/ospf_abr.c:
```diff
@@ -87,13 +87,13 @@ ospf_area_range_add (struct ospf_area *area, struct ospf_area_range *range)
 
 static void
-ospf_area_range_delete (struct ospf_area *area, struct route_node *rn)
+ospf_area_range_delete (struct ospf_area *area, struct route_node *rn, struct ospf *ospf)
```
Let's put `struct ospf *ospf` as the first parameter.

chiragshah6 replied:
I have addressed in latest patch and passed ospf struct as first parameter.

donaldsharp commented on ospfd/ospf_flood.c:
```diff
@@ -83,21 +83,28 @@ ospf_external_info_check (struct ospf_lsa *lsa)
   struct prefix_ipv4 p;
   struct route_node *rn;
   int type;
+  vrf_id_t vrf_id = VRF_DEFAULT;
```
Let's just pass in `struct ospf *ospf` here.

chiragshah6 replied:
I have addressed in latest patch.

donaldsharp commented on ospfd/ospf_opaque.c:
```diff
@@ -560,6 +560,9 @@ register_opaque_info_per_type (struct ospf_opaque_functab *functab,
     case OSPF_OPAQUE_AS_LSA:
+      if (IS_DEBUG_OSPF_TRACE)
+        zlog_debug ("%s: default ospf lookup. vrf_id %u",
+              __PRETTY_FUNCTION__, new->oi->ospf->vrf_id);
```
Let's print the vrf name not the vrf id.  People think about the vrf names not the id.

chiragshah6 replied:
I have addressed in latest patch

donaldsharp commented on ospfd/ospf_opaque.c:
```diff
@@ -560,6 +560,9 @@ register_opaque_info_per_type (struct ospf_opaque_functab *functab,
       listnode_add (new->area->opaque_lsa_self, oipt);
       break;
     case OSPF_OPAQUE_AS_LSA:
+      if (IS_DEBUG_OSPF_TRACE)
```
Why does OSPF_OPAQUE_AS_LSA get a trace but other paths don't?  What special here?

chiragshah6 replied:
this opaue lsa retrieves ospf instace differently then others. The trace is meant for internal daemon activity/probing.  I have addressed in latest patch.

donaldsharp commented on ospfd/ospf_opaque.c:
```diff
@@ -672,6 +675,9 @@ lookup_opaque_info_by_type (struct ospf_lsa *lsa)
           zlog_warn ("Type-11 Opaque-LSA: Reference to OSPF is missing?");
           break; /* Unlikely to happen. */
         }
+      if (IS_DEBUG_OSPF_TRACE)
```
Why does only this case get a trace but other paths don't?  What's special here?  Why do we need this?

chiragshah6 replied:
I have addressed it in latest patch

donaldsharp commented on ospfd/ospf_opaque.c:
```diff
@@ -1598,6 +1604,9 @@ ospf_opaque_lsa_install (struct ospf_lsa *lsa, int rt_recalc)
         }
       break;
     case OSPF_OPAQUE_AS_LSA:
+      if (IS_DEBUG_OSPF_TRACE)
```
Isn't this IS_DEBUG_OSPF (lsa, LSA_INSTALL)?  Why does this case statement get a trace but the others dont?

chiragshah6 replied:
I have addressed it in latest patch

donaldsharp commented on ospfd/ospf_opaque.c:
```diff
@@ -1621,10 +1630,13 @@ ospf_opaque_lsa_install (struct ospf_lsa *lsa, int rt_recalc)
+  if (IS_DEBUG_OSPF_TRACE)
+    zlog_debug ("%s: default ospf lookup. opaque lsa->oi->ospf->vrf_id %u",
+                __PRETTY_FUNCTION__, lsa->oi->ospf->vrf_id);
   ospf = ospf_lookup ();
```
leaving two api's around that do the same thing, one vrf aware and one not is a severe recipe for bugs.  I've already fixed multiple instances of bugs introduced by having a non vrf aware api called when it should have been using the vrf aware.  Just remove ospf_lookup and replace with an appropriate vrf aware function

chiragshah6 replied:
Most places, I have addressed to use vrf aware apis. There are places still required scoping to make vrf aware, snmp, ospf ri etc. Currently primary functionality is working with VRF aware. I will address them time to come ...

donaldsharp commented on ospfd/ospf_packet.c:
```diff
@@ -891,6 +891,17 @@ ospf_hello (struct ip *iph, struct ospf_header *ospfh,
   p.prefixlen = ip_masklen (hello->network_mask);
   p.u.prefix4 = iph->ip_src;
 
+  if (IS_DEBUG_OSPF_TRACE)
```
Why is this not IS_DEBUG_OSP_PACKET?

chiragshah6 replied:
he purpose of the log is for daemon internal activity and provides different filter level of logs. If any objection will remove it.

donaldsharp commented on ospfd/ospf_packet.c:
```diff
@@ -3409,6 +3420,13 @@ ospf_hello_send_sub (struct ospf_interface *oi, in_addr_t addr)
 
   op->dst.s_addr = addr;
 
+  if (IS_DEBUG_OSPF_TRACE)
```
Why is this new debug not a packet debug?

chiragshah6 replied:
The purpose of the log is for daemon internal activity and provides different filter level of logs. If any objection will remove it.

donaldsharp commented on ospfd/ospf_spf.c:
```diff
@@ -1336,6 +1336,10 @@ ospf_spf_calculate_timer (struct thread *thread)
 
   /* Update routing table. */
   monotime(&start_time);
+  if (IS_DEBUG_OSPF_TRACE)
```
Why is this an trace instead of an event?

chiragshah6 replied:
The purpose of the log is for daemon internal activity.

donaldsharp commented on ospfd/ospf_te.c:
```diff
@@ -1244,7 +1244,10 @@ ospf_mpls_te_lsa_new (struct ospf_area *area, struct mpls_te_link *lp)
 
-      struct ospf *top = ospf_lookup ();
+      struct ospf *top = area->ospf;
+      if (OSPF_DEBUG_TRACE)
```
Why is this not IS_DEBUG_OSPF (lsa, LSA_GENERATE)?

chiragshah6 replied:
The purpose of the log is for daemon internal activity. Similar to debug pim trace or trace details

donaldsharp commented on ospfd/ospf_te.c:
```diff
@@ -1551,6 +1554,10 @@ ospf_mpls_te_lsa_refresh (struct ospf_lsa *lsa)
   else
     top = ospf_lookup ();
 
+  if (OSPF_DEBUG_TRACE)
```
Why is this debug not IS_DEBUG_OSPF (lsa, LSA_GENERATE) and why not just expand the debug immediately below?

chiragshah6 replied:
The purpose of the log is for daemon internal activity. Started using as part of vrf development. If any objection, I will remove it.

donaldsharp commented on ospfd/ospf_vty.c:
```diff
@@ -139,23 +140,38 @@ ospf_oi_count (struct interface *ifp)
 DEFUN_NOSH (router_ospf,
        router_ospf_cmd,
-       "router ospf [(1-65535)]",
+       "router ospf [(1-65535)] [vrf VRFNAME]",
```
@eqvinox had me put this back to [vrf NAME] so that we would stay a bit more consistent.  We will still get the vrf auto-complete with the way it was changed

chiragshah6 replied:
I have moved all CLIs back to [vrf NAME]

donaldsharp commented on ospfd/ospf_vty.c:
```diff
@@ -139,23 +140,38 @@ ospf_oi_count (struct interface *ifp)
        "Start OSPF configuration\n"
-       "Instance ID\n")
+       "Instance ID\n"
+       "OSPF VRF\n"
```
VRF_CMD_HELP_STR

chiragshah6 replied:
In latest patch using VRF_CMD_HELP_STR for all VRF related commands

donaldsharp commented on ospfd/ospf_vty.c:
```diff
@@ -165,7 +181,14 @@ DEFUN_NOSH (router_ospf,
         zlog_debug ("Config command 'router ospf %d' received", instance);
-      ospf->oi_running = 1;
+
+      if (ospf->vrf_id != VRF_UNKNOWN)
```
Why not just modify the IS_DEBUG_OSPF_EVENT immediately above to have the right information?

chiragshah6 replied:
In latest patch combined into OSPF_EVENT

donaldsharp commented on ospfd/ospf_vty.c:
```diff
@@ -564,6 +622,10 @@ DEFUN (no_ospf_network_area,
   VTY_GET_IPV4_PREFIX ("network prefix", p, argv[idx_ipv4_prefixlen]->arg);
   VTY_GET_OSPF_AREA_ID (area_id, format, argv[idx_ipv4_number]->arg);
 
+  if (IS_DEBUG_OSPF_TRACE)
```
This new debug gives us no context of what we are processing?  What does this actually help us do?

chiragshah6 replied:
This was for internal daemon activity log, function name print provides context to what cli is executed with parameters

donaldsharp commented on ospfd/ospf_vty.c:
```diff
@@ -526,6 +580,10 @@ DEFUN (ospf_network_area,
   VTY_GET_IPV4_PREFIX ("network prefix", p, argv[idx_ipv4_prefixlen]->arg);
   VTY_GET_OSPF_AREA_ID (area_id, format, argv[idx_ipv4_number]->arg);
 
+  if (IS_DEBUG_OSPF_TRACE)
```
This new debug gives us no context of what we are processing.  What does this actually help us do?

donaldsharp commented on ospfd/ospf_vty.c:
```diff
@@ -3977,11 +4239,27 @@ DEFUN (show_ip_ospf_neighbor_int,
+      if (!ospf->oi_running)
+        continue;
+      if (IS_DEBUG_OSPF_TRACE)
+        zlog_debug ("%s: ifp->vrf_id %d ospf->vrf_id %d", __PRETTY_FUNCTION__,
```
This debug gives us no clue as to what command is being run.  What actual debugability does this provide us?

chiragshah6 replied:
I have removed in latest patch but it was internal daemon activity log.

donaldsharp commented on ospfd/ospf_vty.c:
```diff
@@ -6368,9 +6818,12 @@ DEFUN (no_ip_ospf_dead_interval,
-      if ((ospf = ospf_lookup()))
+      struct ospf *ospf = ospf_lookup_by_vrf_id (ifp->vrf_id);
+
+      if (IS_DEBUG_OSPF_TRACE)
```
I'm going to stop looking at every new debug in this file, but if they must provide a person coming through later the ability to figure out what is going on else they are useless.  What are we providing that is useful long term?

chiragshah6 replied:
I have removed it in latest patch. The debug_trace is meant to use for internal daemon activity/probing logs. function name printed in logs provides which cli is executed with parameters.

donaldsharp commented on ospfd/ospf_vty.c:
```diff
@@ -7909,32 +8375,30 @@ show_ip_ospf_route_network (struct vty *vty, struct route_table *rt)
   struct listnode *node;
   struct ospf_path *path;
 
-  vty_out (vty, "============ OSPF router routing table =============%s",
```
Where did this display line go?

chiragshah6 replied:
The display is moved under per vrf. It is present in diff.

donaldsharp commented on ospfd/ospf_zebra.c:
```diff
@@ -66,23 +66,44 @@ static int
+  ospf = ospf_lookup_by_vrf_id (vrf_id);
+  if (!ospf)
+    {
+      if (IS_DEBUG_OSPF_TRACE)
```
Shouldn't this debug be IS_DEBUG_OSPF (ZEBRA, ZEBRA_INTERFACE)?

chiragshah6 replied:
The OSPF_TRACE is for internal debug activity wherever appropriate also used ZEBRA_INTERFACE

donaldsharp commented on ospfd/ospf_zebra.c:
```diff
@@ -341,9 +371,35 @@ ospf_interface_link_params (int command, struct zclient *zclient,
+  if_update (ifp, ifp->name, strlen (ifp->name), new_vrf_id);
+
+  ospf = ospf_lookup_by_vrf_id (new_vrf_id);
+  if (!ospf)
```
We don't need this if statement

chiragshah6 replied:
I have addressed in new patch.

donaldsharp commented on ospfd/ospf_zebra.c:
```diff
@@ -352,12 +408,18 @@ ospf_zebra_add (struct prefix_ipv4 *p, struct ospf_route *or)
+
+  if (!ospf)
+    {
+      if (OSPF_DEBUG_TRACE)
```
I can't believe this trace here is going to provide any information?

chiragshah6 replied:
The purpose of the trace is for internal daemon activity. Please check new patch.

donaldsharp commented on ospfd/ospf_zebra.c:
```diff
@@ -1089,10 +1168,14 @@ ospf_zebra_read_ipv4 (int command, struct zclient *zclient,
   if (ospf == NULL)
     return 0;
 
+  if (IS_DEBUG_OSPF_TRACE)
```
IS_DEBUG_OSPF( zebra, ZEBRA_REDISTRIBUTE)?

chiragshah6 replied:
This is meant for OSPF daemon internal debug activity. If any objections, I will remove it.

donaldsharp commented on ospfd/ospfd.c:
```diff
@@ -115,6 +115,10 @@ ospf_router_id_update (struct ospf *ospf)
   else
     router_id = router_id_zebra;
 
+  if (IS_DEBUG_OSPF_TRACE)
```
Why is this not IS_DEBUG_OSPF_EVENT?

chiragshah6 replied:
this is meant for OSPF daemon internal debug activity.

chiragshah6 commented:
In latest patch, I have addressed followings

1. Most places instead of using ospf_lookup() , using vrf_id specific or passing ospf as input.

2.  DEBUG_TRACE is developed to use for daemon internal activity  just like debug pim trace or debug pim trace details.  OSPF vrf patch is using in context of VRF development but scope is not limited to VRF only. 
 Removed DEBUG_TRACE or converted them to OSPF_EVENT or others related in critical places. 

3. ospf_spf.c has 2 VRF_DEFAULTS are not active. I do not have full context yet, i will address to remove them in future if required.

4. show ip ospf mpls-te interface  is made vrf aware

5. Tested some of the primary OSPF functionalities with VRF, verified show commands for output. New command show ip ospf vrfs provides configured VRF aware OSPF with vrf_ids and router_ids.

eqvinox commented:
As far as I understand it, this does 2 things:
* adds VRF tagging onto an OSPF instance
* adds multi-instance same-process support

Considering we already have multi-instance support in OSPF, wouldn't it be easier to do VRFs in separate processes? It certainly would provide fault/crash isolation...

chiragshah6 commented:
Hello David,

Current OSPF supports  Multi-Instance or single OSPF Instance (daemon(s)).
The VRF support is made both Multi-Instance and default no instance fall
backs to instance 0 (single daemon).

We thought of doing VRF as separate daemon and running in one of the
daemon/process but to support MI-OSPF with VRF,
hierarchy supports both instance and vrf.

Mostly tested with single instance VRF and also ensured Mi-OSPF and default
behavior is intact.

Hope address the concern.

Kind Regards,
Chirag

On Sat, Jun 17, 2017 at 1:50 PM, David Lamparter <notifications@github.com>
wrote:

> As far as I understand it, this does 2 things:
>
>    - adds VRF tagging onto an OSPF instance
>    - adds multi-instance same-process support
>
> Considering we already have multi-instance support in OSPF, wouldn't it be
> easier to do VRFs in separate processes? It certainly would provide
> fault/crash isolation...
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/FRRouting/frr/pull/734#issuecomment-309239442>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ATqeedCtDxQd7UBhZ77GjW_1d4pID1cNks5sFDwRgaJpZM4N8_m6>
> .
>

vivek-cumulus commented:
Hi David, VRF does bring with it notions of route leaking etc. This can of course still be done across processes, but following the model of how it was done in BGP, we decided to introduce the VRF construct within OSPF. MI-OSPF by itself is to just keep the database contained and it was easy to implement that with distinct OSPF daemons. There is the possibility of running multiple instances of OSPF, each implementing a set of VRFs. There are vendors supporting this too. The changes in this PR should support that too in future.

odd22 commented on ospfd/ospf_te.c:
```diff
@@ -1244,7 +1244,10 @@ ospf_mpls_te_lsa_new (struct ospf_area *area, struct mpls_te_link *lp)
       lsa_id.s_addr = htonl (tmp);
 
-      struct ospf *top = ospf_lookup ();
+      struct ospf *top = area->ospf;
```
Even if struct ospf could be retrieve from area variable, in the case of INTER_AS LSA flood in the AS (LSA Type 11), the area input variable is not defined i.e. could be null. So, it is not safer to get the struct ospf pointer from area->ospf. It is why I use ospf_lookup() call instead. Otherwise, area is defined and could be used.

So, it must be change to:
struct ospf *top = ospf_lookup() or like appear at line 1552 by:

if (area)
   top = area->ospf;
else
    top = ospf_lookup ();
But, this imply to declare struct ospf *top before

odd22 replied:
After more investigation, there is many call to ospf_lookup() function in ospf_te.c, ospf_ri.c and ospf_opaque.c to manage LSA Type 11 with AS flooding scope. All these instance must be replace with new VRF API. But, the problem is to get the vrf_id.

It needs deep investigation in OPAQUE LSA management to determine if the vrf_id could be accessible or need to be pass to these functions.

In this state, this PR could not be merge without breaking OPAQUE LSA Type 11 flooding.

chiragshah6 commented:
@odd22 
Hi Olivier,
I have incorporated all your review comments and created macro for old ospf_lookup() to vrf base api. Reverted to places you have mentioned.

Regards,
Chirag

donaldsharp commented on ospfd/ospf_abr.c:
```diff
@@ -87,13 +87,13 @@ ospf_area_range_add (struct ospf_area *area, struct ospf_area_range *range)
   if (range->specifics != 0)
     ospf_delete_discard_route (area->ospf->new_table,
-			       (struct prefix_ipv4 *) &rn->p);
+			       (struct prefix_ipv4 *) &rn->p, ospf);
```
ospf needs to be first parameter

donaldsharp commented on ospfd/ospf_abr.c:
```diff
@@ -268,7 +268,7 @@ ospf_area_range_unset (struct ospf *ospf, struct in_addr area_id,
     ospf_schedule_abr_task (ospf);
 
-  ospf_area_range_delete (area, rn);
+  ospf_area_range_delete (area, rn, ospf);
```
ospf needs to be first parameter

donaldsharp commented on ospfd/ospf_abr.c:
```diff
@@ -1706,10 +1706,10 @@ ospf_abr_manage_discard_routes (struct ospf *ospf)
 	    if (range->specifics)
 	      ospf_add_discard_route (ospf->new_table, area,
-				      (struct prefix_ipv4 *) &rn->p);
+				      (struct prefix_ipv4 *) &rn->p, ospf);
```
For every single function that you have added ospf to the parameter list, please make it the first.

donaldsharp commented on ospfd/ospf_dump.c:
```diff
@@ -1503,6 +1505,33 @@ DEFUN (no_debug_ospf_te,
   return CMD_SUCCESS;
 }
 
+DEFUN (debug_ospf_trace,
```
I specifically asked for this to be rethought and put into something like debug ospf vrf.

chiragshah6 replied:
This is meant for OSPF daemon internal debug activity. This trace filter not just meant for vrf.
 Similar to debug pim trace.

chiragshah6 commented:
Closing this PR as required to accomodate further review comments to have 
1) to handle ospf TE for vrf
2) make 'struct ospf *' the first function parameter
3) There merge conflicts PR 764 and vtysh related PRs. 

After making modification including all recent changes from top of master will submit new PR.
